### ISSUE TITLE
Add exchanges

### ISSUE THREAD

**Issue by remipaeta** [cryptoeax/arbbot#5]
Is it possible to add exchanges ? I can also pay you privately.

**Comment by remipaeta** [comment]
I tried to add a third exchange but I have errors on Arbitrator.php

**Comment by kingchenx** [comment]
Its better u wait for the rework code :D 
And yea, me too interested to add more exchanges

**Comment by cryptoeax** [comment]
Which exchanges are you interested in @rpcrypto?

**Comment by kingchenx** [comment]
Hitbtc, kraken, yobit, bitfinex and so... many mores :P

**Comment by cryptoeax** [comment]
Haha, well sure, but to be honest I didn't really intend to start maintaining this open source project, this sort of happened as a result of the original project getting blocked.  :-)

I would be happy to do some work on it if there is a sensible way to fund the development but otherwise my contributions so far was more scratching my own itch...  Adding support for exchanges is a lot of work.

**Comment by cryptoeax** [comment]
If people have any particular exchange they would like to see supported, please feel free to comment here.  I'm working on some future improvements including support for some new exchanges.

**Comment by remipaeta** [comment]
I would like Bleutrade, Cryptopia and Binance

**Comment by Scorpic1** [comment]
Please add support Binance, Cryptopia and Yobit.
Thanks.

**Comment by ericfraselle** [comment]
Is it possible to add Bitfitnex, Bitlish, Kraken and GDAX?

**Comment by cryptoeax** [comment]
Why GDAX?  It only supports a few markets so I don't believe supporting it would be that beneficial...

**Comment by ericfraselle** [comment]
You're right

**Comment by onesource** [comment]
I would like to see another exchange like Binance added so that I can remove Poloniex before they require the verification of legacy accounts, which requires a SSN.

**Comment by ericfraselle** [comment]
For me, I can't try the bot because it's no more possible to signup on bittrex for the moment. It's why I would like to add some other exchange.
To replace an exchange is it the poloniex.php and/or bittrex.php, the only file(s) to swap with another exchange API?

**Comment by cryptoeax** [comment]
> For me, I can't try the bot because it's no more possible to signup on bittrex for the moment. It's why I would like to add some other exchange.

Yep, I'm very aware of this problem.  :-/  Have you seen #54?

> To replace an exchange is it the poloniex.php and/or bittrex.php, the only file(s) to swap with another exchange API?

Mostly yes.  You also need to provide `bot/xchange/map.ID` and `web/xchange/ID.ico` as well.  You can use the [bleutrade branch](https://github.com/cryptoeax/arbbot/tree/bleutrade) as an example.  I'd be happy to help answer questions etc. in case you're interested to contribute it back.  :-)

**Comment by beannt** [comment]
I love to have support for coinexchange.io and cryptopia, since there are many pairs with high spread.

**Comment by ElSaico** [comment]
The Cryptopia API seems to be fairly straightforward - quite similar to Bittrex, despite not being <s>a blatant ripoff</s> very similar like Bleutrade.
 
https://www.cryptopia.co.nz/Forum/Thread/255 - Public API
https://www.cryptopia.co.nz/Forum/Thread/256 - Private API

Maybe I'll get my hands dirty on this; no promises though, since they are quite full these days.

**Comment by ElSaico** [comment]
Sometimes I really wish the exchanges had sandbox APIs - there's even testing blockchains for BTC and BCH, so why not that?

**Comment by ElSaico** [comment]
Well, turns out I've started working on Cryptopia support - anyone who wants to take a peek at my code is welcome to do so:

https://github.com/ElSaico/arbbot/blob/cryptopia/bot/xchange/Cryptopia.php

Things missing so far:

- About half the class methods (i.e. the actually complicated ones 😛)
- Importing order history from CSV file

**Comment by cryptoeax** [comment]
> I love to have support for coinexchange.io and cryptopia, since there are many pairs with high spread.

@rpcrypto is already working on Cryptopia (sadly it is no longer possible to register there and I don't have an account there myself so won't even be able to verify the code myself).  It would be nice if you would help with coinexchange.io!  If you'd like to work on it, you can use exchange ID 8.  If you need help feel free to ask questions either here or on gitter.  Thank you!

**Comment by beannt** [comment]
Sure, I will give a try on coinexchange.io. Since PHP is not my main programming language, it's might take some time, but I willing to contribute.

**Comment by cryptoeax** [comment]
> The Cryptopia API seems to be fairly straightforward - quite similar to Bittrex, despite not being a blatant ripoff very similar like Bleutrade.
> 
> https://www.cryptopia.co.nz/Forum/Thread/255 - Public API
> https://www.cryptopia.co.nz/Forum/Thread/256 - Private API

It would be nice to share code with the `Bleutrade` class as much as possible, the same way it shares code with `Bittrex` right now with the `BittrexLikeExchange`.  You may for example be able to inherit from `Bleutrade` and customize methods like `getID()`, `getName()` and so on.

BTW a more like explanation behind exchanges ripping each other off is they buying the same exchange software and customizing it.  :-)  I doubt each exchange writes their own from scratch.



> Sometimes I really wish the exchanges had sandbox APIs - there's even testing blockchains for BTC and BCH, so why not that?

In my experience they for the most part don't have working APIs even... Each one has various bugs that I have had to work around so far, so we are far from that ideal world.  :-(

Longer term I think https://github.com/ccxt/ccxt would be a better solution.  But longer longer term I have plans to redo the whole bot from scratch so I'm not sure how much time I'd spend on porting existing exchanges to use ccxt...

**Comment by cryptoeax** [comment]
> Sure, I will give a try on coinexchange.io. Since PHP is not my main programming language, it's might take some time, but I willing to contributes.

Thank you, and no worries, PHP isn't my main language either, but it's quite easy to pick up and has great documentation on php.net

One question though.  Is this all of the API support they have?  http://coinexchangeio.github.io/slate/  I'm afraid this isn't nearly enough, they offer no wallet management or trading APIs...

**Comment by beannt** [comment]
Just looked at it already. They haven't supported trading api yet. I think I will find another good exchange instead. Skipping coinexchange.io for now, until they release v2 api.

**Comment by cryptoeax** [comment]
Another option is (if you're up for it) to look at their website with your browser's devtools and poke into their JS code to see what it does.  I have done this for Poloniex recently with some success but it obviously takes more time!

**Comment by beannt** [comment]
That's not bad idea. The hardest thing to do with browser way is dealing with authentication/cookie/captcha/2FA things, might need some time to analyze the method they use , any suggestion? since you known how to work with poloniex before.

**Comment by beannt** [comment]
I found another exchange instead, tradesatoshi, found some open source on github https://github.com/WorldBot/Trade-Satoshi-API, a good way to start :)

**Comment by cryptoeax** [comment]
> That's not bad idea. The hardest thing to do with browser way is dealing with authentication/cookie/captcha/2FA things, might need some time to analyze the method they use , any suggestion? since you known how to work with poloniex before.

For Poloniex I was trying to construct real time orderbook data from their websocket API (the code isn't in this repo) and was just reading their code to see how it processes the incoming data since their API is undocumented.  But the auth issue is a big one, typically exchanges either have an API that uses some kind of HMAC based authentication where you digitally sign each request with your API key (see any of our exchange implementations for examples) or they use a cookie that the exchange sign-in system sets when you use the browser to sign in, and APIs that are hidden behind the latter form of authentication aren't useful for automation through bots very easily.

That being said, these days you can run Firefox and Chrome in headless mode, and there are libraries like https://github.com/GoogleChrome/puppeteer that allow you to automate the browser, so one could potentially look into adding the functionality to log in using puppeteer to the exchange and lift the cookie and give it to the bot, but that would require the account to not have 2FA enabled which is quite insecure.  I'm not inclined to accept such changes for security reasons... :)

**Comment by cryptoeax** [comment]
> I found another exchange instead, tradesatoshi, found some open source on github https://github.com/WorldBot/Trade-Satoshi-API, a good way to start :)

OK cool.  It's sad that exchange also doesn't have registrations open any more!

**Comment by cryptoeax** [comment]
A couple of recent Exchange API changes were pushed today (https://github.com/cryptoeax/arbbot/commit/b72135eadb6d0fa74d743e7fec528e00776e0419 and https://github.com/cryptoeax/arbbot/commit/04b1fd587774b6abed31a2ccdf74b8860c7efcdc). This is of interest to those who are implementing new exchanges.  You are welcome to ignore the newly added arguments like the existing exchanges do.

**Comment by beannt** [comment]
That's unlucky, many exchanges closed registration recently. I think I will try binance or kucoin since they have just opened limited registration each day.

**Comment by Intel1146** [comment]
hitbtc would be awesome, thanks

**Comment by thosuperman** [comment]
I think binance would be a good choice, https://github.com/binance-exchange/binance-official-api-docs

**Comment by cryptoeax** [comment]
> I think binance would be a good choice, https://github.com/binance-exchange/binance-official-api-docs

I'm working on an implementation of Binance on top of [ccxt](https://github.com/ccxt/ccxt).  :-)

**Comment by beannt** [comment]
Nice, many supported exchanges with ccxt. It's seem the right way to do. Some info: Cryptopia have just opened for registration :-)

**Comment by cryptoeax** [comment]
Yes, I think at this point it's better to wait for my CCXTAdapter class and use it for further exchanges, it makes adding new exchanges much simpler!

**Comment by cryptoeax** [comment]
I posted the binance branch here: https://github.com/cryptoeax/arbbot/issues/109#issuecomment-358747572  Those who want to use the `CCXTAdapter` class can start doing so now!

**Comment by sinaxi** [comment]
what exchanges are supported now and works (trades) without any problems?
Thanks!
